backend/src/stwo/prover.rs

```diff
@@ -1,95 +1,227 @@
 use powdr_ast::analyzed::Analyzed;
+use powdr_backend_utils::machine_fixed_columns;
+use powdr_executor::constant_evaluator::VariablySizedColumn;
+use powdr_number::FieldElement;
 use serde::de::DeserializeOwned;
 use serde::ser::Serialize;
+use std::collections::BTreeMap;
 use std::io;
 use std::marker::PhantomData;
 use std::sync::Arc;
 
 use crate::stwo::circuit_builder::{gen_stwo_circuit_trace, PowdrComponent, PowdrEval};
+use crate::stwo::proof::{StarkProvingKey, TableProvingKey, TableProvingKeyCollection};
 
-use stwo_prover::constraint_framework::TraceLocationAllocator;
+use stwo_prover::constraint_framework::{
+    TraceLocationAllocator, ORIGINAL_TRACE_IDX, PREPROCESSED_TRACE_IDX,
+};
 use stwo_prover::core::prover::StarkProof;
 
-use powdr_number::FieldElement;
 use stwo_prover::core::air::{Component, ComponentProver};
-use stwo_prover::core::backend::{Backend, BackendForChannel};
+use stwo_prover::core::backend::{Backend, BackendForChannel, Column, ColumnOps};
 use stwo_prover::core::channel::{Channel, MerkleChannel};
-use stwo_prover::core::fields::m31::M31;
+use stwo_prover::core::fields::m31::{BaseField, M31};
 use stwo_prover::core::fri::FriConfig;
 use stwo_prover::core::pcs::{CommitmentSchemeProver, CommitmentSchemeVerifier, PcsConfig};
-use stwo_prover::core::poly::circle::CanonicCoset;
+use stwo_prover::core::poly::circle::{CanonicCoset, CircleEvaluation};
+use stwo_prover::core::poly::twiddles::TwiddleTree;
+use stwo_prover::core::poly::BitReversedOrder;
+use stwo_prover::core::utils::{bit_reverse_index, coset_index_to_circle_domain_index};
+use stwo_prover::core::ColumnVec;
 
 const FRI_LOG_BLOWUP: usize = 1;
 const FRI_NUM_QUERIES: usize = 100;
 const FRI_PROOF_OF_WORK_BITS: usize = 16;
 const LOG_LAST_LAYER_DEGREE_BOUND: usize = 0;
 
-pub struct StwoProver<T, B: Backend + Send, MC: MerkleChannel, C: Channel> {
+pub struct StwoProver<T, B: BackendForChannel<MC> + Send, MC: MerkleChannel, C: Channel> {
     pub analyzed: Arc<Analyzed<T>>,
-    _fixed: Arc<Vec<(String, Vec<T>)>>,
-    /// Proving key placeholder
-    _proving_key: Option<()>,
+    /// The split analyzed PIL
+    split: BTreeMap<String, Analyzed<T>>,
+    /// The value of the fixed columns
+    pub fixed: Arc<Vec<(String, VariablySizedColumn<T>)>>,
+
+    /// Proving key
+    proving_key: StarkProvingKey<B, MC>,
     /// Verifying key placeholder
     _verifying_key: Option<()>,
     _channel_marker: PhantomData<C>,
-    _backend_marker: PhantomData<B>,
-    _merkle_channel_marker: PhantomData<MC>,
 }
 
 impl<'a, F: FieldElement, B, MC, C> StwoProver<F, B, MC, C>
 where
-    B: Backend + Send + BackendForChannel<MC>, // Ensure B implements BackendForChannel<MC>
+    B: Backend + Send + BackendForChannel<MC>,
```

Comment:
Right, we have a `Send` bound here, so `TableProvingKey<B, MC>` should be automatically `Send` as well. Maybe there's a bound missing somewhere?

Reply:
Reference: https://doc.rust-lang.org/nomicon/send-and-sync.html